File: templates/nodejs-service/service/src/transport/interfaces.ts
export type Message = { [key: string]: any };
export type RPCHandler = (...inputs: any[]) => any;
export type EventHandler = (input: Message) => void;

export interface RPCClient {
    call: (functionName: string, ...inputs: any[]) => Promise<any>;
}

export interface RPCServer {
    registerHandler: (functionName: string, handler: RPCHandler) => RPCServer;
    serve: () => Promise<void>;
}

export interface Publisher {
    publish: (eventName: string, msg: Message) => Promise<void>;
}

export interface Subscriber {
    registerHandler: (eventName: string, handler: EventHandler) => Subscriber;
    subscribe: () => Promise<void>;
}

export interface RmqEventConfig {
    queueName: string;
    exchangeName: string;
    rpcTimeout?: number;
    deadLetterExchange: string;
    deadLetterQueue: string;
    ttl: number;
    autoAck: boolean;
}

File: templates/go-service/service/transport/interfaces.go
package transport

// RPCHandler function to handle things
type RPCHandler func(inputs ...interface{}) (output interface{}, err error)

// EventHandler function to handle things
type EventHandler func(input Message) (err error)

// RPCClient interface
type RPCClient interface {
	Call(functionName string, inputs ...interface{}) (output interface{}, err error)
}

// RPCServer interface
type RPCServer interface {
	RegisterHandler(functionName string, handler RPCHandler) RPCServer
	Serve(errChan chan error)
}

// Publisher interface
type Publisher interface {
	Publish(functionName string, msg Message) (err error)
}

// Subscriber interface
type Subscriber interface {
	RegisterHandler(functionName string, handler EventHandler) Subscriber
	Subscribe(errChan chan error)
}
